UX-1661 - Add API Docs to Dev Center

langlk opened the pull request:
Resolves dev-center side of [UX-1661](https://algorithmia.atlassian.net/browse/FRONTEND-1661)

langlk commented on server/index.js:
```diff
@@ -33,41 +33,72 @@ app.use((req, res, next) => {
 })
 
-// Remove trailing slashes
+// API Docs - resolve before trailing slash redirect so assets don't break
```
Unfortunately it's proving extremely frustrating to get the api-docs to work in Express without the trailing slash. We end up with styles broken on either the Express server version, or the `localhost:4567` version.

bnlb commented on README.md:
```diff
@@ -52,39 +52,30 @@ gem install --install-dir vendor/bundle/ruby/<ruby_version_numbwe> nokogiri -v "
- 6. Run `git submodule update`, cloning down submodule code.
- 7. To build Synapse dist files, run `cd synapse && npm ci && npm run build && cd ..`
- 8. Start the local server:
+ 5. To set up submodules and install dependencies, run `npm run setup`.
```
Can we add a note about making sure that `rvm` is installed? I got an error when running `npm run setup` because I didn't have `nvm` on my machine. https://rvm.io/rvm/install

langlk replied:
Added!

bnlb commented:
I haven't looked into this just yet, but I got a few errors while running `npm install`. Curious if you've seen them before:
```
❯ npm install

> gifsicle@0.1.7 postinstall /Users/nlb/work/dev-center/node_modules/gifsicle
> node index.js

path.js:39
    throw new ERR_INVALID_ARG_TYPE('path', 'string', path);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type object
    at assertPath (path.js:39:11)
    at Object.basename (path.js:1294:5)
    at /Users/nlb/work/dev-center/node_modules/download/index.js:35:43
    at each (/Users/nlb/work/dev-center/node_modules/each-async/each-async.js:63:4)
    at module.exports (/Users/nlb/work/dev-center/node_modules/download/index.js:33:5)
    at /Users/nlb/work/dev-center/node_modules/bin-wrapper/index.js:108:20
    at /Users/nlb/work/dev-center/node_modules/bin-wrapper/index.js:141:24
    at /Users/nlb/work/dev-center/node_modules/bin-check/index.js:30:20
    at /Users/nlb/work/dev-center/node_modules/executable/index.js:39:20
    at FSReqWrap.oncomplete (fs.js:154:21)

> optipng-bin@0.3.11 postinstall /Users/nlb/work/dev-center/node_modules/optipng-bin
> node index.js

path.js:39
    throw new ERR_INVALID_ARG_TYPE('path', 'string', path);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type object
    at assertPath (path.js:39:11)
    at Object.basename (path.js:1294:5)
    at /Users/nlb/work/dev-center/node_modules/download/index.js:35:43
    at each (/Users/nlb/work/dev-center/node_modules/each-async/each-async.js:63:4)
    at module.exports (/Users/nlb/work/dev-center/node_modules/download/index.js:33:5)
    at /Users/nlb/work/dev-center/node_modules/bin-wrapper/index.js:108:20
    at /Users/nlb/work/dev-center/node_modules/bin-wrapper/index.js:141:24
    at /Users/nlb/work/dev-center/node_modules/bin-check/index.js:30:20
    at /Users/nlb/work/dev-center/node_modules/executable/index.js:39:20
    at FSReqWrap.oncomplete (fs.js:154:21)
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: gifsicle@0.1.7 (node_modules/gifsicle):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: gifsicle@0.1.7 postinstall: `node index.js`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: optipng-bin@0.3.11 (node_modules/optipng-bin):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: optipng-bin@0.3.11 postinstall: `node index.js`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1
```

langlk commented on package.json:
```diff
@@ -7,21 +7,33 @@
-    "start": "NODE_ENV=production node ./server/index.js",
+    "apidocs:build": "cd api-docs && bundle exec rake build",
+    "apidocs:setup": "git submodule init && git submodule update && cd api-docs && rvm install 2.3.0 && gem install bundle && bundle install",
+    "dev": "(npm run apidocs:build && npm run synapse:build && npm run devcenter:dev) & npm server:dev",
```
Fixed!

langlk commented:
@bnlb I've seen those errors, but since they're optional dependencies and don't seem to be causing any issues with running the app, they don't seem super high-priority.

bnlb commented:
> @bnlb I've seen those errors, but since they're optional dependencies and don't seem to be causing any issues with running the app, they don't seem super high-priority.

Good to know, just wanted to check!

bnlb commented on package.json:
```diff
@@ -7,21 +7,33 @@
   "scripts": {
-    "start": "NODE_ENV=production node ./server/index.js",
+    "apidocs:build": "cd api-docs && bundle exec rake build",
+    "apidocs:setup": "git submodule init && git submodule update && cd api-docs && rvm install 2.3.0 && gem install bundle && bundle install",
```
I got an error on this command when getting set up:
```
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems.rb:241:in `bin_path': can't find gem bundler (>= 0.a) (Gem::GemNotFoundException)
        from /usr/local/bin/bundle:22:in `<main>'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! algorithmia-dev-center@1.0.0 apidocs:setup: `git submodule init && git submodule update && cd api-docs && rvm install 2.3.0 && gem install bundle && bundle install`
```
What's interesting is that running each command individually worked fine.

langlk replied:
Agh, for some reason I'm not able to repro 😬are you still seeing that when you run the `setup` script?

bnlb commented on server/index.js:
```diff
@@ -33,41 +33,72 @@ app.use((req, res, next) => {
+    target: config.env.stage.devCenterUrl,
+    changeOrigin: true,
+  }
+  app.use(require('http-proxy-middleware')(devCenterProxyConfig))
```
This is actually my fault, as I hadn't yet updated `package.json` when I handed this branch off to you, but I got an error related to `http-proxy-middleware` not being installed when I ran `npm run dev`. Turns out I forgot to add it to package.json so it was never installed.

langlk replied:
Added!

bnlb commented:
> @bnlb I've seen those errors, but since they're optional dependencies and don't seem to be causing any issues with running the app, they don't seem super high-priority.

Fwiw, updating `grunt-contrib-imagemin` to `"^3.1.0"` resolves this error. I looked through the releases in that repo and the only potentially breaking changes I see are them dropping support for Node v4. Best to leave as-is for now until we can do a more thorough check, just wanted to mention.

bnlb commented on package.json:
```diff
@@ -7,21 +7,33 @@
   "license": "The MIT License (MIT)",
   "scripts": {
-    "start": "NODE_ENV=production node ./server/index.js",
+    "apidocs:build": "cd api-docs && bundle exec rake build",
```
I had to update this line to 
```
"apidocs:build": "cd api-docs && bundle install && bundle exec rake build",
```
in order to get it to work.

langlk replied:
Okay, will update!

bnlb commented:
My only other thought is that it could be worth updating the port numbers we use for these different servers. We currently use the following:

• node server: 3000
• local dev-center Jekyll server: 4000

In order to see the api docs listed at /developers/api, we have to use the Node server to serve the prebuilt content via Express.static. This makes `npm run dev` more useful for local development than running `bundle exec jekyll serve`. When running `npm run localdevcenter` in Golduck, we hit port 4000 by default when we proxy requests to dev-center. I vote we change the default port used in the express server to 4000 and bump the Jekyll server to 4001.

Adding @peckjon since he'll be using these commands more than either of us.

andryanmiller commented on server/index.js:
```diff
@@ -33,41 +33,72 @@ app.use((req, res, next) => {
-  if (/.+\/$/.test(req.path)) {
+  if (
+    hasTrailingSlash(req.path) &&
+    !(process.env.LOCAL || isApiDocs(req.path))
```
:bulb: Might we be able to consolidate on `NODE_ENV` for this? Might be nice to save ourselves the extra env var.

langlk replied:
I think that's a good call. Will update!

langlk commented:
This is deployed to test right now, but the web-server changes have not been deployed to test yet, so `/developers/api` redirect is still infinite

bnlb commented:
I ended up getting the scripts all working. It was a combination of needing to reinstall xcode developer tools for Catalina, a bunch of Grunt dependencies we weren't using that complicated things, and the fact that we were calling `gem install bundle` instead of `gem install bundler`. Super easy to miss that `r`.

I also removed the `npm run dev` command because I ran into an issue while testing this where after I'd control-c the terminal window the Node server would close but the hot reloading Jekyll server would still be running and be unclose-able. I've updated the docs to reflect that.

langlk commented:
> I ended up getting the scripts all working. It was a combination of needing to reinstall xcode developer tools for Catalina, a bunch of Grunt dependencies we weren't using that complicated things, and the fact that we were calling `gem install bundle` instead of `gem install bundler`. Super easy to miss that `r`.
> 
> I also removed the `npm run dev` command because I ran into an issue while testing this where after I'd control-c the terminal window the Node server would close but the hot reloading Jekyll server would still be running and be unclose-able. I've updated the docs to reflect that.

Thanks! I'm making a couple last changes and then this should finally be ready to go!

bnlb commented on package.json:
```diff
@@ -7,11 +7,23 @@
     "e2e:tunnel": "E2E_BASE_URL=http://localhost:3000 E2E_USE_TUNNEL=true npm run e2e:cloud",
     "e2e:local": "E2E_BASE_URL=http://localhost:3000 npm run e2e",
-    "start": "NODE_ENV=production node ./server/index.js",
+    "setup": "npm run apidocs:setup && npm run apidocs:build && npm run synapse:setup && gem install bundler && bundle install && npm install",
```
How come we have to run `apidocs:build` as part of the setup script?

langlk replied:
So that we have the `build/` directory set up so the express server can find those files.

I figured since the api-docs repo doesn't get updated much, making sure those files were built during setup was generally more convenient than adding an extra step.
